### PR TITLE
Search 183 clone - Metrics to IResearchLink

### PR DESCRIPTION
**Pull request by alexbakharew**
### Scope & Purpose

*(Please describe the changes in this PR for reviewers - **mandatory**)*

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [X] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [X] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: arangod/IResearch/IResearchLink.cpp [hunk 1]
```diff
@@ -969,6 +978,22 @@ Result IResearchLink::drop() {
...
 
+  std::string viewId, colId, shardName, dbName;
+
+  getLinkLabels(viewId, colId, shardName, dbName);
+
+  auto tmpStats = arangodb_arangosearch_link_stats{};
+  tmpStats.addLabel("viewId", viewId);
+  tmpStats.addLabel("collId", colId);
+  tmpStats.addLabel("shardName", shardName);
+  tmpStats.addLabel("dbName", dbName);
```
Comment (alexbakharew) [inline arangod/IResearch/IResearchLink.cpp:989]: Move it to getLinkLabels function

File: arangod/IResearch/IResearchLink.cpp [hunk 2]
```diff
@@ -1234,6 +1259,20 @@ Result IResearchLink::init(
...
+  std::string viewID, colId, shard, dbName;
+
+  getLinkLabels(viewID, colId, shard, dbName);
+
+  auto tmpStats = arangodb_arangosearch_link_stats{};
+  tmpStats.addLabel("viewId", viewId);
+  tmpStats.addLabel("collId", colId);
+  tmpStats.addLabel("shardName", shard);
+  tmpStats.addLabel("dbName", dbName);
+
```
Comment (alexbakharew) [inline arangod/IResearch/IResearchLink.cpp:1271]: Move it to getLinkLabels function

File: arangod/IResearch/IResearchLink.cpp [hunk 3]
```diff
@@ -1872,6 +1927,22 @@ Result IResearchLink::unload() {
...
+  std::string viewId, colId, shardName, dbName;
+
+  getLinkLabels(viewId, colId, shardName, dbName);
+
+  auto tmpStats = arangodb_arangosearch_link_stats{};
+  tmpStats.addLabel("viewId", viewId);
+  tmpStats.addLabel("collId", colId);
+  tmpStats.addLabel("shardName", shardName);
+  tmpStats.addLabel("dbName", dbName);
+
```
Comment (alexbakharew) [inline arangod/IResearch/IResearchLink.cpp:1939]: Move it to getLinkLabels function

File: tests/IResearch/IResearchLink-test.cpp [hunk 4]
```diff
@@ -864,6 +869,332 @@ TEST_F(IResearchLinkTest, test_write) {
...
+                        size_t& indexSize,
+                        size_t& numFiles) {
+
+  const std::regex regex("^_\\d+$");
+  std::smatch match;
+
+  for (auto& p: fs::directory_iterator(path)) {
+    std::string name = p.path().stem().string();
+    std::string ext = p.path().extension().string();
+
```
Comment (alexbakharew) [inline tests/IResearch/IResearchLink-test.cpp:883]: use irs::utf8_path

File: tests/IResearch/IResearchLink-test.cpp [hunk 5]
```diff
@@ -24,6 +24,9 @@
```
Comment (alexbakharew) [inline tests/IResearch/IResearchLink-test.cpp:24]: Add tests for links creation, for metric output

File: arangod/IResearch/IResearchLink.cpp [hunk 6]
```diff
@@ -1522,6 +1561,22 @@ void IResearchLink::scheduleConsolidation(std::chrono::milliseconds delay) {
...
+  viewId = _viewGuid;
+  dbName = _collection.vocbase().id();
+
+  if (ServerState::instance()->isDBServer()) {
+    shardName = _collection.name();
+    colId = _meta._collectionName;
+  } else if (ServerState::instance()->isSingleServer()) {
+    colId = std::to_string(_collection.id().id());
+    shardName = "";
+  }
```
Comment (gnusi) [inline arangod/IResearch/IResearchLink.cpp:1577]:
```suggestion
} else {
  TRI_ASSERT(false);
}
```

File: tests/IResearch/IResearchLink-test.cpp (hunk 4 above)
Comment (alexbakharew) [reply tests/IResearch/IResearchLink-test.cpp]: remove all std::filesystem entries

**Comment by alexbakharew** [comment]
Work in this PR currently postponed.
At this moment, I have following problem: launching unittests using "../scripts/unittest dump --cluster true"
cause segfaults. This problem comes from "init()" function, where you create a metric for current link. After using couts in each "init()" function it is evidently that it tries to create link with same labels. Need to debug this problem more precisely

**Comment by alexbakharew** [comment]
Also it is necessary to add another bunch of tests. The scenarios should come from problems which was encountered in previous debugging session. For example:

1. create link
2. check link stats
3. try to create link with same parameters. Ensure that this will cause exception
4. unload 1st link
5. init 2nd link
6. check link stats

**Comment by MBkkt** [comment]
Closed by this https://github.com/arangodb/arangodb/pull/15075

**Pull request closed without merging**